influxql/functions.go

```diff
@@ -757,3 +757,137 @@ func (r *FloatHoltWintersReducer) constrain(x []float64) {
 		x[3] = 0
 	}
 }
+
+// FloatIntegralReducer calculates the time-integral of the aggregated points.
+type FloatIntegralReducer struct {
+	groupByTime bool
+	count       uint32
+	interval    Interval
+	sum         float64
+	prev        FloatPoint
+}
+
+// NewFloatIntegralReducer creates a new FloatIntegralReducer.
+func NewFloatIntegralReducer(interval Interval, groupByTime bool) *FloatIntegralReducer {
+	return &FloatIntegralReducer{
+		groupByTime: groupByTime,
+		interval:    interval,
+		prev:        FloatPoint{Nil: true},
+	}
+}
+
+// AggregateFloat aggregates a point into the reducer.
+func (r *FloatIntegralReducer) AggregateFloat(p *FloatPoint) {
+
+	// If this is the first point, just save it
+	if r.prev.Nil {
+		r.prev = *p
+		r.count++
+		return
+	}
+
+	// If this point has the same timestamp as the previous one,
+	// use it but don't add anything to the integral.  Effectively
+	// this means that we allow the curve we are integrating
+	// to have step changes in it, but who knows whether the points
+	// actually arrive in any particular order...?
+	if r.prev.Time == p.Time {
+		r.prev = *p
+		r.count++
+		return
+	}
+
+	// Normal operation: update the sum using the trapezium rule
+	elapsed := float64(p.Time-r.prev.Time) / float64(r.interval.Duration)
+	r.sum += 0.5 * (p.Value + r.prev.Value) * elapsed
+	r.prev = *p
+	r.count++
+	return
+}
+
+// Emit emits the time-integral of the aggregated points as a single point.
+// InfluxQL convention dictates that outside a group-by-time clause we return
+// a timestamp of zero.  Within a group-by-time, we can set the time to ZeroTime
+// and a higher level will change it to the start of the time group.
+func (r *FloatIntegralReducer) Emit() []FloatPoint {
+	if r.groupByTime {
```

Comment:
This shouldn't be needed. If you use `ZeroTime`, one of the other iterators takes care of setting the start time correctly. Using `0` would result in the wrong time here.